Feat(eos_cli_config_gen, eos_designs): Flooded Traffic using Multicast Underlay

Vibhu-gslab opened the pull request:
## Change Summary

Support for underlay multicast replication of flooded traffic in L2 EVPN domains

## Related Issue(s)

Fixes #2033 

## Component(s) name

`arista.avd.eos_designs`
`arista.avd.eos_cli_config_gen`

## Proposed changes
Adding model in cli_config_gen and eos_designs

## How to test
CI

## Checklist

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

Vibhu-gslab committed:
Feat(eos_cli_config_gen, eos_designs): Flooded Traffic using Multicast Underlay

Vibhu-gslab committed:
Fixing CI

Vibhu-gslab committed:
Adding eos_design part

ClausHolbechArista commented on ansible_collections/arista/avd/examples/cv-pathfinder/documentation/devices/site1-border1.md:
```diff
@@ -584,10 +584,10 @@ interface Vlan4094
-| ---- | --- | ---------- | --------------- |
-| 42 | 10042 | - | - |
-| 666 | 10666 | - | - |
+| VLAN | VNI | Flood List | Flood Group Multicast IP | Multicast Group |
```
Instead of adding a field here, I think we should just fill out the IP under the `Flood List` column.

Vibhu-gslab replied:
Fixed!

laxmikantchintakindi replied:
Can we show this IP as <ip_address>(multicast)
e.g 239.1.1.1(multicast)

gmuloc replied:
(multicast) not needed

ClausHolbechArista commented on python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/vxlan_interface.schema.yml:
```diff
@@ -119,6 +119,9 @@ keys:
                       items:
                         type: str
                         description: Remote VTEP IP Address.
+                    flood_group_multicast_ipaddress:
```
This is not similar to the CLI.
```suggestion
                    flood_group:
```

Vibhu-gslab replied:
Fixed!

ClausHolbechArista commented on python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/vxlan_interface.schema.yml:
```diff
@@ -119,6 +119,9 @@ keys:
                         description: Remote VTEP IP Address.
+                    flood_group_multicast_ipaddress:
+                      type: str
+                      description: Underlay IP multicast group address.
```
```suggestion
                      description: Flood multicast group address.
```

Vibhu-gslab replied:
Fixed!

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -191,6 +191,15 @@ $defs:
               default: 0
               convert_types:
                 - str
+            underlay_l2_multicast_flood_group_ipv4_pool:
```
This is not bound to EVPN L2 multicast. It should be in a separate dict like `vxlan_flood_multicast` with it's own `enabled` key and `underlay_multicast_group_ipv4_pool`, `underlay_multicast_group_ipv4_pool_offset`. On individual vlans it should be the same but just with `enabled` and `underlay_multicast_group` keys.

Vibhu-gslab replied:
Fixed!

ClausHolbechArista reviewed (changes requested):
Please rename wording in code and descriptions accordingly.

Vibhu-gslab committed:
Fixing cli_config_gen model

Vibhu-gslab committed:
fixing eos_designs model

Vibhu-gslab committed:
Fixing eos_designs conditions and adding tests

laxmikantchintakindi commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -197,6 +196,22 @@ def _set_vxlan_interface_config_for_vlan(
                 vlan.id,
                 tenant.evpn_l2_multicast.underlay_l2_multicast_group_ipv4_pool_offset,
             )
+        if vlan.vxlan_flood_multicast.enabled is not None:
```
`if vlan.vxlan_flood_multicast.enabled:`
if we only check it is enabled then code below this line can be optimized.

laxmikantchintakindi commented on python-avd/pyavd/api/ip_addressing/__init__.py:
```diff
@@ -337,6 +337,16 @@ def evpn_underlay_l2_multicast_group(
         offset = vlan_id - 1 + underlay_l2_multicast_group_ipv4_pool_offset
         return get_ip_from_pool(underlay_l2_multicast_group_ipv4_pool, 32, offset, 0)
 
+    def evpn_underlay_l2_flood_group(
```
May be rename this to evpn_underlay_l2_multicast_flood_group or evpn_underlay_l2_flood_multicast_group for clarity?

Vibhu-gslab committed:
Improving coverage

Vibhu-gslab committed:
Fixing CI

jonxstill reviewed (commented):
I think this is mostly fine - however should we be checking if `underlay_multicast` is enabled to make sure we have some kind of multicast underlay transport configured before we start setting up multicast flood groups?

ClausHolbechArista commented:
> I think this is mostly fine - however should we be checking if `underlay_multicast` is enabled to make sure we have some kind of multicast underlay transport configured before we start setting up multicast flood groups?

Good point. Let's add that @Vibhu-gslab. This may also require an update to the other PR where multicast is being changed. This should only work with multicast_pim_sm.

Vibhu-gslab committed:
adding underlay_multicast check

Vibhu-gslab committed:
fixing CI

laxmikantchintakindi commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -197,6 +196,23 @@ def _set_vxlan_interface_config_for_vlan(
                 vlan.id,
                 tenant.evpn_l2_multicast.underlay_l2_multicast_group_ipv4_pool_offset,
             )
+        if self.shared_utils.underlay_multicast:
```
This should be reversed. Flood group should be added when underlay_multicast is not configured.

Vibhu-gslab committed:
adding underlay multicast check and testcase

Vibhu-gslab commented:
> > I think this is mostly fine - however should we be checking if `underlay_multicast` is enabled to make sure we have some kind of multicast underlay transport configured before we start setting up multicast flood groups?
> 
> Good point. Let's add that @Vibhu-gslab. This may also require an update to the other PR where multicast is being changed. This should only work with multicast_pim_sm.

Added the check.

laxmikantchintakindi commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -198,6 +197,25 @@ def _set_vxlan_interface_config_for_vlan(
                 tenant.evpn_l2_multicast.underlay_l2_multicast_group_ipv4_pool_offset,
             )
 
+        if not self.shared_utils.underlay_multicast and (vlan.vxlan_flood_multicast.enabled is not None or tenant.vxlan_flood_multicast.enabled is True):
```
```suggestion
        if not self.shared_utils.underlay_multicast and (vlan.vxlan_flood_multicast.enabled or tenant.vxlan_flood_multicast.enabled):
```

Vibhu-gslab committed:
Update python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py

Co-authored-by: laxmikantchintakindi <159624484+laxmikantchintakindi@users.noreply.github.com>

laxmikantchintakindi commented on ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FLOOD_MULTICAST_MISSING_VXLAN_IPV4_POOL.yml:
```diff
@@ -0,0 +1,31 @@
+  node_groups:
+    - group: flood_multicast_missing
+      nodes:
+        - name: flood_multicast_missing_1a
```
```suggestion
        - name: flood-multicast-missing-1a
```

laxmikantchintakindi replied:
Same for all the hostnames

Vibhu-gslab replied:
I think original one is fine.

Vibhu-gslab replied:
Fixed!

Vibhu-gslab committed:
fixing node names

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:
```diff
@@ -128,6 +133,10 @@ tenants:
+    vxlan_flood_multicast:
+      enabled: true
+      underlay_l2_multicast_group_ipv4_pool: 232.1.0.0-232.1.0.40
+      underlay_l2_multicast_group_ipv4_pool_offset: 1
```
Maybe we should make the default value of the offset 1, so we align the group address with the vlan number by default.

Vibhu-gslab replied:
We have default set as `0` for `evpn_l2_multicast.underlay_l2_multicast_group_ipv4_pool_offset`
https://github.com/aristanetworks/avd/blob/devel/python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml#L189

It might get bit confusing for the user.

ClausHolbechArista replied:
agreed. Keep what you have.

gmuloc replied:
we can move default value to 1 for both in 6.0 lets add a TODO

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml:
```diff
@@ -160,6 +160,17 @@ $defs:
+          table: network-services-multicast-settings
+        type: dict
+        description: |
+          Explicitly enable or disable vxlan_flood_multicast to override setting of `<network_services_key>.[].vxlan_flood_multicast.enabled`.
```
```suggestion
          Explicitly enable or disable vxlan_flood_multicast to override settings of `<network_services_key>.[].vxlan_flood_multicast`.
```

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -198,6 +197,25 @@ def _set_vxlan_interface_config_for_vlan(
+        if not self.shared_utils.underlay_multicast and (vlan.vxlan_flood_multicast.enabled or tenant.vxlan_flood_multicast.enabled):
+            msg = "'vxlan_flood_multicast' is only supported in combination with 'underlay_multicast: True'."
+            raise AristaAvdError(msg)
+        if vlan.vxlan_flood_multicast.enabled is not None:
+            vxlan_vlan.flood_group = (
+                vlan.vxlan_flood_multicast.underlay_multicast_group
+                if vlan.vxlan_flood_multicast.underlay_multicast_group and vlan.vxlan_flood_multicast.enabled is True
+                else None
+            )
+        elif tenant.vxlan_flood_multicast.enabled is True:
+            if not tenant.vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool:
+                msg = f"'vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool' for Tenant: {tenant.name} is required."
+                raise AristaAvdInvalidInputsError(msg)
+            vxlan_vlan.flood_group = self.shared_utils.ip_addressing.evpn_underlay_l2_flood_group(
+                tenant.vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool,
```
I am not sure this will work well. What happens if I enable under the vlan but wish to rely on the pool to get the group? So not setting enabled to false under the tenant and overriding with true under the vlan.
I think you need to set a vxlan_vlan_flood_multicast_enabled using the default() between the two booleans. And then if the result is true you can look for the direct ip under the vlan or fall back to the pool from the tenant level.

Vibhu-gslab replied:
Changed the condition as asked, just did not use `vxlan_vlan_flood_multicast_enabled` variable.
Added a testcase as well.

Vibhu-gslab committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml

Co-authored-by: Claus Holbech <holbech@arista.com>

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Vibhu-gslab committed:
Adding a testcase and changing condition

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

laxmikantchintakindi commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -198,6 +197,25 @@ def _set_vxlan_interface_config_for_vlan(
                 tenant.evpn_l2_multicast.underlay_l2_multicast_group_ipv4_pool_offset,
             )
 
+        if not self.shared_utils.underlay_multicast and (vlan.vxlan_flood_multicast.enabled or tenant.vxlan_flood_multicast.enabled):
+            msg = "'vxlan_flood_multicast' is only supported in combination with 'underlay_multicast: True'."
+            raise AristaAvdError(msg)
+        if default(vlan.vxlan_flood_multicast.enabled, tenant.vxlan_flood_multicast.enabled):
```
```suggestion
        if default(vlan.vxlan_flood_multicast.enabled, tenant.vxlan_flood_multicast.enabled):
            if not self.shared_utils.underlay_multicast:
                msg = "'vxlan_flood_multicast' is only supported in combination with 'underlay_multicast: True'."
                raise AristaAvdError(msg)
```

laxmikantchintakindi replied:
Can this be invalid inputs error?

laxmikantchintakindi commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -198,6 +197,25 @@ def _set_vxlan_interface_config_for_vlan(
+            )
+            if vxlan_vlan.flood_group is None:
+                if not tenant.vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool:
+                    msg = f"'vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool' for Tenant: {tenant.name} is required."
```
```suggestion
                    msg = f"'vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool' for Tenant: {tenant.name} or 'vxlan_flood_multicast.underlay_l2_multicast_group' for VLAN: {vlan.id} is required."
```

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Update python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py

Co-authored-by: laxmikantchintakindi <159624484+laxmikantchintakindi@users.noreply.github.com>

Vibhu-gslab committed:
Update python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py

Co-authored-by: laxmikantchintakindi <159624484+laxmikantchintakindi@users.noreply.github.com>

Vibhu-gslab committed:
fixing testcase and error message

Vibhu-gslab committed:
Adding required: true

laxmikantchintakindi commented on ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FLOOD_MULTICAST_MISSING_VXLAN_IPV4_POOL.yml:
```diff
@@ -0,0 +1,32 @@
+    vtep_loopback_ipv4_pool: 10.43.0.0/24
+    mlag_interfaces: [ Ethernet5, Ethernet6 ]
+    mlag_peer_ipv4_pool: 10.255.252.0/24
+    mlag_peer_l3_ipv4_pool: 10.255.251.0/24
```
```suggestion
```

laxmikantchintakindi commented on ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FLOOD_MULTICAST_MISSING_UNDERLAY_MULTICAST.yml:
```diff
@@ -0,0 +1,30 @@
+    - group: flood_multicast_missing
+      nodes:
+        - name: flood-multicast-missing-underlay-multicast-1a
+          bgp_as: 42
```
These two negative tests can be moved to host vars. As they are for a single host.

Vibhu-gslab replied:
Done!

laxmikantchintakindi commented on ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FLOOD_MULTICAST_MISSING_UNDERLAY_MULTICAST.yml:
```diff
@@ -0,0 +1,30 @@
+  defaults:
+    loopback_ipv4_pool: 10.42.0.0/24
+    vtep_loopback_ipv4_pool: 10.43.0.0/24
+    mlag_interfaces: [ Ethernet5, Ethernet6 ]
+    mlag_peer_ipv4_pool: 10.255.252.0/24
+    mlag_peer_l3_ipv4_pool: 10.255.251.0/24
```
```suggestion
```

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Update ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FLOOD_MULTICAST_MISSING_UNDERLAY_MULTICAST.yml

Co-authored-by: laxmikantchintakindi <159624484+laxmikantchintakindi@users.noreply.github.com>

Vibhu-gslab committed:
Update ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FLOOD_MULTICAST_MISSING_VXLAN_IPV4_POOL.yml

Co-authored-by: laxmikantchintakindi <159624484+laxmikantchintakindi@users.noreply.github.com>

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
moving to hostvars

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

MaheshGSLAB commented on python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/vxlan-interface.j2:
```diff
@@ -64,6 +64,9 @@ interface Vxlan1
 {%         if vlan.flood_vteps is arista.avd.defined %}
    vxlan vlan {{ vlan.id }} flood vtep {{ vlan.flood_vteps | join(' ') }}
 {%         endif %}
+{%         if vlan.flood_group is arista.avd.defined %}
```
The all flood vtep comes first then all flood group then all multicast group.
```
s1-leaf1(config-if-Vx1)#sh ac
interface Vxlan1
   vxlan udp-port 4789
   vxlan vlan 11 flood vtep 20.1.1.1
   vxlan vlan 111 flood vtep 10.1.1.10 10.1.1.11
   vxlan vlan 11 flood group 232.1.1.22
   vxlan vlan 111 flood group 232.1.1.21
   vxlan vlan 11 multicast group 232.1.1.22
   vxlan vlan 110 multicast group 239.9.1.4
```

MaheshGSLAB replied:
Like we are running the loop again for multicast group similarly we need to run separate loop for flood group.

Vibhu-gslab replied:
Thanks.. added!

MaheshGSLAB commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -201,6 +200,29 @@ def _set_vxlan_interface_config_for_vlan(
                 tenant.evpn_l2_multicast.underlay_l2_multicast_group_ipv4_pool_offset,
             )
 
+        if default(vlan.vxlan_flood_multicast.enabled, tenant.vxlan_flood_multicast.enabled):
```
shouldn't we check the  vxlan_flood_multicast.enabled for vrf before to tenant?

MaheshGSLAB commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:
```diff
@@ -224,6 +242,9 @@ tenants:
       - id: 7
         name: "MULTICAST_DISABLED_7"
         tags: ['test_l2']
+        vxlan_flood_multicast:
```
Can you please create a new id may be 10 with description as vxlan_flood_multicast_enabled

MaheshGSLAB replied:
Similar to other places and add the comment about test, like globally or tenant level disabled and vlan level enabled

Vibhu-gslab replied:
Added!

Vibhu-gslab committed:
fixing order

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
modifying testcases and adding comment

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

MaheshGSLAB reviewed (approved):
LGTM

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py:
```diff
@@ -201,6 +200,29 @@ def _set_vxlan_interface_config_for_vlan(
+            if vxlan_vlan.flood_group is None:
+                if not tenant.vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool:
+                    msg = (
+                        "'vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool' for Tenant: "
+                        f"{tenant.name} or 'vxlan_flood_multicast.underlay_l2_multicast_group' for VLAN: "
+                        f"{vlan.id} is required."
```
```suggestion
                        "'vxlan_flood_multicast.underlay_l2_multicast_group_ipv4_pool' for Tenant "
                        f"'{tenant.name}' or 'vxlan_flood_multicast.underlay_l2_multicast_group' for VLAN "
                        f"'{vlan.id}' is required."
```

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1503,6 +1524,18 @@ $defs:
+                  enabled:
+                    type: bool
+                    required: true
+                  underlay_multicast_group:
```
```suggestion
                  underlay_multicast_group:
                    description: Specific multicast group to use for this VLAN.
```

ClausHolbechArista replied:
same on svis

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1503,6 +1524,18 @@ $defs:
+                keys:
+                  enabled:
+                    type: bool
+                    required: true
```
```suggestion
```

ClausHolbechArista replied:
same on svis

Vibhu-gslab replied:
Done!

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -1503,6 +1524,18 @@ $defs:
+                documentation_options:
+                  table: network-services-multicast-settings
+                type: dict
+                description: |
+                  Explicitly enable or disable vxlan_flood_multicast to override setting of `<network_services_key>.[].vxlan_flood_multicast.enabled`.
```
Move this to the enabled key.

ClausHolbechArista replied:
same on svis

Vibhu-gslab replied:
Done!

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -202,6 +203,26 @@ $defs:
+              required: true
+            underlay_l2_multicast_group_ipv4_pool:
+              type: str
+              description: Comma separated list of prefixes (IPv4 address/Mask) or ranges (IPv4_address-IPv4_address).
```
```suggestion
              description: |-
                IPv4 Pool to allocate vxlan flood groups from. 
                The VLAN ID - 1 + `underlay_l2_multicast_group_ipv4_pool_offset` is the offset for each group.
                Comma separated list of prefixes (IPv4_address/Mask) or ranges (IPv4_address-IPv4_address).
```

Vibhu-gslab replied:
Done!

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml:
```diff
@@ -202,6 +203,26 @@ $defs:
+          documentation_options:
+            table: network-services-multicast-settings
+          description: |
+            Enable Flood group Multicast for all SVIs and l2vlans within Tenant.
```
Move this to the enabled key.

Vibhu-gslab replied:
Done!

ClausHolbechArista commented on python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/vxlan-interface.j2:
```diff
@@ -65,6 +65,11 @@ interface Vxlan1
 {%         endif %}
 {%     endfor %}
+{%     for vlan in vxlan_config.vxlan.vlans | arista.avd.natural_sort('id') %}
+{%         if vlan.flood_group is arista.avd.defined %}
```
Please check on EOS if we can have flood vteps on a vlan together with a flood group. If not, we could combine this into the for loop above, to avoid looping twice.

Vibhu-gslab replied:
We can have both `flood vteps` and `flood group` together.

```
interface Vxlan1
   vxlan udp-port 4789
   vxlan vlan 2 flood vtep 1.1.1.1
   vxlan vlan 2 flood group 232.1.1.21
```

ClausHolbechArista commented on python-avd/pyavd/api/ip_addressing/__init__.py:
```diff
@@ -343,6 +343,16 @@ def evpn_underlay_l2_multicast_group(
+        underlay_l2_flood_ipv4_pool_offset: int,
+    ) -> str:
+        """Return IP address to be used for EVPN underlay L2 flood group."""
+        offset = vlan_id - 1 + underlay_l2_flood_ipv4_pool_offset
```
Let's remove the -1 here. So we just take vlan id + offset.

ClausHolbechArista replied:
nvm

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Update python-avd/pyavd/_eos_designs/structured_config/network_services/vxlan_interface.py

Co-authored-by: Claus Holbech <holbech@arista.com>

Vibhu-gslab committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml

Co-authored-by: Claus Holbech <holbech@arista.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Vibhu-gslab committed:
Addressing comments on schema

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Vibhu-gslab committed:
Fixing CI

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

joelbreton2 commented:
When I run the PR against the dual-dc example I receive the following error: 
```
TASK [arista.avd.eos_cli_config_gen : Generate eos intended configuration and device documentation] 
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: jinja2.exceptions.UndefinedError: 'flood_list' is undefined
fatal: [dc1-leaf1b -> localhost]: FAILED! => {"changed": false, "msg": "Error during plugin execution: 'flood_list' is undefined"}
```

Changes in the group_vars
FABRIC.yml
```
underlay_multicast: true
```
NETWORK_SERVICES.yml
```
    l2vlans:
      - id: 3401
        name: L2_VLAN3401
      - id: 3402
        name: L2_VLAN3402
        vxlan_flood_multicast:
          enabled: true
          underlay_multicast_group: 234.0.0.100
```

Vibhu-gslab committed:
Fixing Error

Vibhu-gslab commented:
> When I run the PR against the dual-dc example I receive the following error:
> 
> ```
> TASK [arista.avd.eos_cli_config_gen : Generate eos intended configuration and device documentation] 
> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: jinja2.exceptions.UndefinedError: 'flood_list' is undefined
> fatal: [dc1-leaf1b -> localhost]: FAILED! => {"changed": false, "msg": "Error during plugin execution: 'flood_list' is undefined"}
> ```
> 
> Changes in the group_vars FABRIC.yml
> 
> ```
> underlay_multicast: true
> ```
> 
> NETWORK_SERVICES.yml
> 
> ```
>     l2vlans:
>       - id: 3401
>         name: L2_VLAN3401
>       - id: 3402
>         name: L2_VLAN3402
>         vxlan_flood_multicast:
>           enabled: true
>           underlay_multicast_group: 234.0.0.100
> ```

Thanks for pointing out! 
I have fixed the condition which was leading to the failure, also added a testcase for the same.

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

joelbreton2 reviewed (dismissed):
not working with dual-dc example

gmuloc commented:
> not working with dual-dc example

This is because multicast is not supported by EVPN GW on EOS

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml:
```diff
@@ -164,6 +164,21 @@ $defs:
+        documentation_options:
+          table: network-services-multicast-settings
+        type: dict
+        description: |-
+          Explicitly enable or disable vxlan_flood_multicast to override settings of `<network_services_key>.[].vxlan_flood_multicast`.
```
```suggestion
```

Vibhu-gslab committed:
Merge branch 'devel' into multicast_underlay

ClausHolbechArista reviewed (approved):
LGTM

ClausHolbechArista committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ClausHolbechArista committed:
Merge branch 'devel' into multicast_underlay
